File: currency.py
"""Currency conversion using Frankfurter API"""
from fastapi import APIRouter, Query, HTTPException
import requests
import time
import logging

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/currency", tags=["currency"])

# Rate cache: {from_to: (rate, timestamp)}
RATE_CACHE = {}
CACHE_TTL = 3600  # 1 hour

@router.get("/convert")
async def convert_currency(
    amount: float = Query(..., gt=0),
    from_currency: str = Query(..., regex="^[A-Z]{3}$"),
    to_currency: str = Query(..., regex="^[A-Z]{3}$")
):
    """Convert currency using Frankfurter API with 1-hour caching"""
    if from_currency == to_currency:
        return {
            "amount": amount,
            "from": from_currency,
            "to": to_currency,
            "converted": amount,
            "rate": 1.0,
            "cached": False
        }

    cache_key = f"{from_currency}_{to_currency}"

    # Check cache
    if cache_key in RATE_CACHE:
        rate, timestamp = RATE_CACHE[cache_key]
        if time.time() - timestamp < CACHE_TTL:
            converted = amount * rate
            logger.info(f"Currency conversion (cached): {amount} {from_currency} = {converted} {to_currency}")
            return {
                "amount": amount,
                "from": from_currency,
                "to": to_currency,
                "converted": converted,
                "rate": rate,
                "cached": True
            }

    # Call Frankfurter API
    try:
        url = f"https://api.frankfurter.app/latest?amount={amount}&from={from_currency}&to={to_currency}"
        response = requests.get(url, timeout=10)
        response.raise_for_status()
        data = response.json()

        converted = data['rates'][to_currency]
        rate = converted / amount

        # Cache the rate
        RATE_CACHE[cache_key] = (rate, time.time())

        logger.info(f"Currency conversion (live): {amount} {from_currency} = {converted} {to_currency}")

        return {
            "amount": amount,
            "from": from_currency,
            "to": to_currency,
            "converted": converted,
            "rate": rate,
            "cached": False
        }

    except Exception as e:
        logger.error(f"Currency conversion failed: {e}")
        raise HTTPException(status_code=503, detail=f"Currency conversion unavailable: {str(e)}")

@router.get("/rates")
async def get_rates(base: str = Query("USD", regex="^[A-Z]{3}$")):
    """Get all exchange rates for a base currency"""
    try:
        url = f"https://api.frankfurter.app/latest?from={base}"
        response = requests.get(url, timeout=10)
        response.raise_for_status()
        data = response.json()

        return {
            "base": base,
            "date": data['date'],
            "rates": data['rates']
        }

    except Exception as e:
        logger.error(f"Failed to fetch rates: {e}")
        raise HTTPException(status_code=503, detail=f"Rates unavailable: {str(e)}")

File: currency.ts
export type Currency = 'USD' | 'GBP' | 'EUR' | 'CAD' | 'AUD' | 'INR' | 'JPY' | 'SGD' | 'AED';

export const CURRENCY_SYMBOLS: Record<Currency, string> = {
  'USD': '$',
  'GBP': '£',
  'EUR': '€',
  'CAD': 'C$',
  'AUD': 'A$',
  'INR': '₹',
  'JPY': '¥',
  'SGD': 'S$',
  'AED': 'د.إ',
};

export const CURRENCY_NAMES: Record<Currency, string> = {
  'USD': 'US Dollar',
  'GBP': 'British Pound',
  'EUR': 'Euro',
  'CAD': 'Canadian Dollar',
  'AUD': 'Australian Dollar',
  'INR': 'Indian Rupee',
  'JPY': 'Japanese Yen',
  'SGD': 'Singapore Dollar',
  'AED': 'UAE Dirham',
};

let fxRateCache = new Map<string, { rate: number; timestamp: number }>();
const CACHE_DURATION = 3600000;
let fxApiAvailable = true;

export function isFxApiAvailable(): boolean {
  return fxApiAvailable;
}

export async function getFxRate(from: string, to: string): Promise<number> {
  if (from === to) return 1;

  const cacheKey = `${from}-${to}`;
  const cached = fxRateCache.get(cacheKey);

  if (cached && Date.now() - cached.timestamp < CACHE_DURATION) {
    return cached.rate;
  }

  try {
    const url = `https://api.frankfurter.app/latest?amount=1&from=${from}&to=${to}`;
    const response = await fetch(url);

    if (!response.ok) {
      throw new Error(`Frankfurter API error: ${response.status}`);
    }

    const data = await response.json();
    const rate = data.rates[to];

    if (!rate) {
      throw new Error(`No rate found for ${from} -> ${to}`);
    }

    fxRateCache.set(cacheKey, { rate, timestamp: Date.now() });
    console.log(`[Currency] FX rate ${from}->${to} = ${rate}`);
    fxApiAvailable = true;

    return rate;
  } catch (error) {
    console.warn(`[Currency] Failed to fetch FX rate ${from}->${to}:`, error);
    fxApiAvailable = false;
    return 1;
  }
}

export async function convertAmount(amount: number, from: string, to: string): Promise<number> {
  const rate = await getFxRate(from, to);
  return amount * rate;
}

export function formatMoney(amount: number, currency: string): string {
  return new Intl.NumberFormat('en-US', {
    style: 'currency',
    currency: currency,
    minimumFractionDigits: 2,
    maximumFractionDigits: 2
  }).format(amount);
}

export async function convertCurrency(
  amount: number,
  fromCurrency: Currency,
  toCurrency: Currency
): Promise<number> {
  return convertAmount(amount, fromCurrency, toCurrency);
}

export function convertCurrencySync(amount: number, fromCurrency: Currency, toCurrency: Currency): number {
  if (fromCurrency === toCurrency) return amount;

  const cacheKey = `${fromCurrency}-${toCurrency}`;
  const cached = fxRateCache.get(cacheKey);

  if (cached) {
    return amount * cached.rate;
  }

  return amount;
}

export function formatPrice(amount: number, currency: Currency): string {
  return formatMoney(amount, currency);
}
